File: tools/memory/safe-memory-search.py
#!/usr/bin/env python3
"""Safe memory search with vector->keyword fallback.

Primary path: `openclaw memory search`.
Fallback path: keyword scan over MEMORY.md and memory/*.md when vector search
is unavailable or embedding quota errors occur.
"""

from __future__ import annotations

import argparse
import json
import os
import re
import subprocess
from pathlib import Path
from typing import Any

WORKSPACE = Path("/Users/hd/openclaw")
STATE_PATH = WORKSPACE / "memory" / "vector-memory-health-state.json"


def load_state() -> dict[str, Any]:
    try:
        return json.loads(STATE_PATH.read_text(encoding="utf-8"))
    except Exception:
        return {}


def save_state(state: dict[str, Any]) -> None:
    STATE_PATH.parent.mkdir(parents=True, exist_ok=True)
    STATE_PATH.write_text(json.dumps(state, indent=2), encoding="utf-8")


def parse_json(raw: str) -> Any | None:
    raw = raw.strip()
    if not raw:
        return None
    m = re.search(r"[\[{]", raw)
    if m:
        raw = raw[m.start() :]
    try:
        return json.loads(raw)
    except Exception:
        return None


def vector_search(query: str, max_results: int) -> tuple[list[dict[str, Any]] | None, str]:
    proc = subprocess.run(
        ["openclaw", "memory", "search", query, "--json", "--max-results", str(max_results)],
        capture_output=True,
        text=True,
        timeout=90,
    )
    combined = f"{proc.stdout}\n{proc.stderr}".strip()
    payload = parse_json(proc.stdout)
    items: list[dict[str, Any]] = []
    if isinstance(payload, list):
        items = [x for x in payload if isinstance(x, dict)]
    elif isinstance(payload, dict):
        for key in ("results", "items", "matches", "data"):
            val = payload.get(key)
            if isinstance(val, list):
                items = [x for x in val if isinstance(x, dict)]
                break

    text_lower = combined.lower()
    quota_error = bool(re.search(r"(resource_exhausted|embedd\w*[^\n]{0,80}429|429[^\n]{0,80}embedd\w*)", text_lower)) and (
        "failed" in text_lower or "error" in text_lower or "quota" in text_lower
    )

    if proc.returncode != 0 or quota_error:
        return None, combined
    return items, combined


def collect_files(workspace: Path) -> list[Path]:
    files: list[Path] = []
    root = workspace / "MEMORY.md"
    if root.exists():
        files.append(root)
    mem = workspace / "memory"
    if mem.exists():
        files.extend(sorted(mem.glob("*.md")))
    return files


def score_line(query_terms: list[str], line: str) -> int:
    l = line.lower()
    score = 0
    for t in query_terms:
        if t in l:
            score += 2
    if len(query_terms) >= 2 and all(t in l for t in query_terms[:2]):
        score += 1
    return score


def keyword_fallback(query: str, max_results: int) -> list[dict[str, Any]]:
    terms = [t for t in re.findall(r"[a-zA-Z0-9_]+", query.lower()) if len(t) >= 3]
    if not terms:
        terms = [query.lower()]

    hits: list[dict[str, Any]] = []
    for p in collect_files(WORKSPACE):
        try:
            lines = p.read_text(encoding="utf-8", errors="ignore").splitlines()
        except Exception:
            continue
        rel = str(p.relative_to(WORKSPACE))
        for idx, line in enumerate(lines, start=1):
            s = score_line(terms, line)
            if s <= 0:
                continue
            snippet = line.strip()
            if len(snippet) > 420:
                snippet = snippet[:417] + "..."
            hits.append(
                {
                    "source": rel,
                    "line": idx,
                    "snippet": snippet,
                    "score": float(s),
                    "mode": "keyword_fallback",
                }
            )

    hits.sort(key=lambda x: x["score"], reverse=True)
    return hits[:max_results]


def main() -> int:
    ap = argparse.ArgumentParser(description="Safe memory search with fallback")
    ap.add_argument("query")
    ap.add_argument("--max-results", type=int, default=5)
    ap.add_argument("--json", action="store_true")
    args = ap.parse_args()

    state = load_state()
    force_fallback = bool(state.get("fallback_mode", False))

    items: list[dict[str, Any]]
    mode = "vector"
    error_text = ""

    if not force_fallback:
        vector_items, error_text = vector_search(args.query, args.max_results)
        if vector_items is not None:
            items = vector_items
        else:
            mode = "keyword_fallback"
            items = keyword_fallback(args.query, args.max_results)
    else:
        mode = "keyword_fallback"
        items = keyword_fallback(args.query, args.max_results)

    if mode == "keyword_fallback":
        state["fallback_mode"] = True
        state["last_fallback_reason"] = error_text[:500]
        save_state(state)

    output = {"mode": mode, "query": args.query, "results": items}
    if args.json:
        print(json.dumps(output, ensure_ascii=False))
    else:
        print(json.dumps(output, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/memory/safe-memory-search.ts
#!/usr/bin/env npx tsx

import fs from "fs";
import path from "path";
import { spawnSync } from "child_process";
import { resolveRepoPath } from "../lib/paths.js";
import { readJsonFile, writeJsonFileAtomic } from "../lib/json-file.js";

const WORKSPACE = resolveRepoPath();
const STATE_PATH = path.join(WORKSPACE, "memory", "vector-memory-health-state.json");

type JsonValue = any;

function loadState(): Record<string, any> {
  const data = readJsonFile<Record<string, any>>(STATE_PATH);
  return data ?? {};
}

function saveState(state: Record<string, any>): void {
  writeJsonFileAtomic(STATE_PATH, state, 2);
}

function parseJson(raw: string): JsonValue | null {
  const text = raw.trim();
  if (!text) return null;
  const m = text.search(/[\[{]/);
  const payload = m >= 0 ? text.slice(m) : text;
  try {
    return JSON.parse(payload);
  } catch {
    return null;
  }
}

function vectorSearch(query: string, maxResults: number): [Array<Record<string, any>> | null, string] {
  const proc = spawnSync("openclaw", ["memory", "search", query, "--json", "--max-results", String(maxResults)], {
    encoding: "utf8",
    timeout: 90000,
  });
  const combined = `${proc.stdout || ""}\n${proc.stderr || ""}`.trim();
  const payload = parseJson(proc.stdout || "");
  let items: Array<Record<string, any>> = [];
  if (Array.isArray(payload)) {
    items = payload.filter((x) => x && typeof x === "object");
  } else if (payload && typeof payload === "object") {
    for (const key of ["results", "items", "matches", "data"]) {
      const val = (payload as any)[key];
      if (Array.isArray(val)) {
        items = val.filter((x) => x && typeof x === "object");
        break;
      }
    }
  }

  const textLower = combined.toLowerCase();
  const quotaError =
    /(resource_exhausted|embedd\w*[^\n]{0,80}429|429[^\n]{0,80}embedd\w*)/.test(textLower) &&
    (textLower.includes("failed") || textLower.includes("error") || textLower.includes("quota"));

  if (proc.status !== 0 || quotaError) {
    return [null, combined];
  }
  return [items, combined];
}

function collectFiles(workspace: string): string[] {
  const files: string[] = [];
  const root = path.join(workspace, "MEMORY.md");
  if (fs.existsSync(root)) files.push(root);
  const memDir = path.join(workspace, "memory");
  if (fs.existsSync(memDir)) {
    for (const p of fs.readdirSync(memDir).sort()) {
      if (p.endsWith(".md")) files.push(path.join(memDir, p));
    }
  }
  return files;
}

function scoreLine(queryTerms: string[], line: string): number {
  const l = line.toLowerCase();
  let score = 0;
  for (const t of queryTerms) {
    if (l.includes(t)) score += 2;
  }
  if (queryTerms.length >= 2 && queryTerms.slice(0, 2).every((t) => l.includes(t))) {
    score += 1;
  }
  return score;
}

function keywordFallback(query: string, maxResults: number): Array<Record<string, any>> {
  let terms = (query.toLowerCase().match(/[a-zA-Z0-9_]+/g) ?? []).filter((t) => t.length >= 3);
  if (!terms.length) terms = [query.toLowerCase()];

  const hits: Array<Record<string, any>> = [];
  for (const p of collectFiles(WORKSPACE)) {
    let lines: string[] = [];
    try {
      lines = fs.readFileSync(p, "utf8").split(/\r?\n/);
    } catch {
      continue;
    }
    const rel = path.relative(WORKSPACE, p);
    lines.forEach((line, idx) => {
      const s = scoreLine(terms, line);
      if (s <= 0) return;
      let snippet = line.trim();
      if (snippet.length > 420) snippet = `${snippet.slice(0, 417)}...`;
      hits.push({
        source: rel,
        line: idx + 1,
        snippet,
        score: Number(s),
        mode: "keyword_fallback",
      });
    });
  }

  hits.sort((a, b) => Number(b.score) - Number(a.score));
  return hits.slice(0, maxResults);
}

function printHelp(): void {
  const text = `usage: safe-memory-search.ts [-h] [--max-results MAX_RESULTS] [--json] query\n\nSafe memory search with fallback\n\npositional arguments:\n  query\n\noptions:\n  -h, --help            show this help message and exit\n  --max-results MAX_RESULTS\n  --json`;
  console.log(text);
}

async function main(): Promise<void> {
  const argv = process.argv.slice(2);
  if (argv.includes("-h") || argv.includes("--help") || argv.length === 0) {
    printHelp();
    if (argv.length === 0) process.exit(2);
    return;
  }

  let maxResults = 5;
  let jsonFlag = false;
  const positional: string[] = [];

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const next = argv[i + 1];
    if (arg === "--max-results" && next) {
      maxResults = Number.parseInt(next, 10);
      i += 1;
    } else if (arg === "--json") {
      jsonFlag = true;
    } else if (arg.startsWith("-")) {
      console.error(`Unknown argument: ${arg}`);
      printHelp();
      process.exit(2);
    } else {
      positional.push(arg);
    }
  }

  const query = positional.join(" ").trim();
  if (!query) {
    console.error("query is required");
    process.exit(2);
  }

  const state = loadState();
  const forceFallback = Boolean(state.fallback_mode);

  let items: Array<Record<string, any>> = [];
  let mode = "vector";
  let errorText = "";

  if (!forceFallback) {
    const [vectorItems, err] = vectorSearch(query, maxResults);
    errorText = err;
    if (vectorItems !== null) {
      items = vectorItems;
    } else {
      mode = "keyword_fallback";
      items = keywordFallback(query, maxResults);
    }
  } else {
    mode = "keyword_fallback";
    items = keywordFallback(query, maxResults);
  }

  if (mode === "keyword_fallback") {
    state.fallback_mode = true;
    state.last_fallback_reason = errorText.slice(0, 500);
    saveState(state);
  }

  const output = { mode, query, results: items };
  if (jsonFlag) {
    console.log(JSON.stringify(output));
  } else {
    console.log(JSON.stringify(output, null, 2));
  }
}

main().catch((err) => {
  const msg = err instanceof Error ? err.message : String(err);
  console.error(msg);
  process.exit(1);
});
